source/logrepl/cdc_test.go

```diff
@@ -485,6 +485,132 @@ func TestCDCIterator_Ack(t *testing.T) {
 		})
 	}
 }
+func TestCDCIterator_NextN(t *testing.T) {
+	ctx := test.Context(t)
+	pool := test.ConnectPool(ctx, t, test.RepmgrConnString)
+	table := test.SetupTestTable(ctx, t, pool)
+
+	t.Run("retrieve exact N records", func(t *testing.T) {
+		is := is.New(t)
+		i := testCDCIterator(ctx, t, pool, table, true)
+		<-i.sub.Ready()
+
+		for j := 1; j <= 3; j++ {
+			_, err := pool.Exec(ctx, fmt.Sprintf(`INSERT INTO %s (id, column1, column2, column3, column4, column5)
+				VALUES (%d, 'test-%d', %d, false, 12.3, 14)`, table, j+10, j, j*100))
+			is.NoErr(err)
+		}
+
+		var allRecords []opencdc.Record
+
+		// Collect records until we have all 3
+		for len(allRecords) < 3 {
+			attemptCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			records, err := i.NextN(attemptCtx, 3-len(allRecords))
+			cancel()
+			is.NoErr(err)
+			// Only proceed if we got at least one record
+			is.True(len(records) > 0)
+			allRecords = append(allRecords, records...)
+		}
+
+		is.Equal(len(allRecords), 3)
+
+		for j, r := range allRecords {
+			is.Equal(r.Operation, opencdc.OperationCreate)
+			is.Equal(r.Key.(opencdc.StructuredData)["id"], int64(j+11))
+			change := r.Payload
+			data := change.After.(opencdc.StructuredData)
+			is.Equal(data["column1"], fmt.Sprintf("test-%d", j+1))
+			//nolint:gosec // no risk to overflow
+			is.Equal(data["column2"], (int32(j)+1)*100)
+		}
+	})
+
+	t.Run("retrieve fewer records than requested", func(t *testing.T) {
+		is := is.New(t)
+		i := testCDCIterator(ctx, t, pool, table, true)
+		<-i.sub.Ready()
+
+		for j := 1; j <= 2; j++ {
+			_, err := pool.Exec(ctx, fmt.Sprintf(`INSERT INTO %s (id, column1, column2, column3, column4, column5)
+				VALUES (%d, 'test-%d', %d, false, 12.3, 14)`, table, j+20, j, j*100))
+			is.NoErr(err)
+		}
+
+		// This requests more than available
+		records, err := i.NextN(ctx, 5)
+		is.NoErr(err)
+
+		// Verify we got fewer records than requested
+		// and that they are in order
+		is.True(len(records) > 0 && len(records) <= 2)
```

Comment:
The assertion made me look twice. So it can happen that the iterator returns either 1 or 2 records, right? I'd change the test to call `NextN` again until we have 2 records. And, additionally, I'd issue _another_ call to `NextN` where we don't expect anything to get back (we can create a context deadline in that case to make sure we don't wait forever). Note that I didn't check if there is maybe already a test that tests what happens if `NextN` doesn't have anything to fetch - if we do, we don't have to do the last call of course.

Along these lines:

```suggestion
		// As long as we don't have 2 records, keep fetching records
		var records []opencdc.Record
		for len(records) < 2 {
			recordsTmp, err := i.NextN(ctx, 5)
			is.NoErr(err)
			records = append(records, recordsTmp...)
		}
		ctxWithTimeout, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
		defer cancel()
		_, err = i.NextN(ctxWithTimeout, 5)
		is.True(errors.Is(err, context.DeadlineExceeded))
```

Reply:
This is sensible